File: apps/data-engine/src/infrastructure/state/checkpoint-reader.ts
import { readFileSync, existsSync } from "fs"

export class CheckpointReadError extends Error {
  constructor(msg: string) {
    super(msg)
    this.name = "CheckpointReadError"
  }
}

export class CheckpointReader {
  constructor(private readonly filePath: string) {}

  exists(): boolean {
    return existsSync(this.filePath)
  }

  readSync<T>(): T {
    if (!this.exists()) {
      throw new CheckpointReadError(`checkpoint not found: ${this.filePath}`)
    }
    try {
      const raw = readFileSync(this.filePath, "utf8")
      return JSON.parse(raw) as T
    } catch (err) {
      throw new CheckpointReadError(
        `failed to read checkpoint: ${err instanceof Error ? err.message : String(err)}`,
      )
    }
  }

  readWithFallback<T>(fallback: T): T {
    if (!this.exists()) return fallback
    try {
      return this.readSync<T>()
    } catch {
      return fallback
    }
  }
}

File: apps/analytics-engine/app/infrastructure/state/checkpoint_reader.py
"""Checkpoint reader — loads and validates a checkpoint from disk."""
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, TypeVar

log = logging.getLogger(__name__)


class CheckpointReadError(RuntimeError):
    pass


T = TypeVar("T")


class CheckpointReader:
    """Reads a checkpoint file with schema validation and fallback."""

    def __init__(self, file_path: str) -> None:
        self._path = Path(file_path)

    def exists(self) -> bool:
        return self._path.exists()

    def read(self) -> dict[str, Any]:
        if not self.exists():
            raise CheckpointReadError(f"checkpoint not found: {self._path}")
        try:
            raw = self._path.read_text(encoding="utf-8")
            data: dict[str, Any] = json.loads(raw)
            return data
        except (json.JSONDecodeError, OSError) as exc:
            raise CheckpointReadError(
                f"failed to read checkpoint: {exc}"
            ) from exc

    def read_with_fallback(self, fallback: dict[str, Any]) -> dict[str, Any]:
        if not self.exists():
            return fallback
        try:
            return self.read()
        except CheckpointReadError:
            return fallback
